File: go/integrations.go
// Package realmid — cross-realm integrations (ADR-082 / ADR-083).
//
// A SOURCE platform publishes an integration; a TARGET org installs it,
// admitting a kind=service principal into the org that holds a chosen
// service-typed role; the source platform then MINTS short-lived target-realm
// access tokens against the installation. GitHub-App-shaped: register once,
// install per org. RI hosts no consent screen — this surface IS the consent
// surface (ADR-083 §5).
//
// The SDK is per-realm: register/mint run on the SOURCE realm's client, and
// install/uninstall run on the TARGET realm's client.
package realmid

import (
	ctxpkg "context"
	"errors"
	"net/url"
	"strconv"
)

// Integration is a source platform's published declaration.
type Integration struct {
	ID          string `json:"id"`
	RealmID     string `json:"realm_id"`
	Slug        string `json:"slug"`
	DisplayName string `json:"display_name"`
	Description string `json:"description"`
	HomepageURL string `json:"homepage_url"`
	// Listed opts into a future discovery directory (ADR-083 §6.1); the listing
	// API is deferred, so today it is metadata only.
	Listed    bool   `json:"listed"`
	Disabled  bool   `json:"disabled"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

// Installation is the target org's view of one inbound edge — who can act in
// the org, as what. Returned by ListInstallations (ADR-083 §4.5).
type Installation struct {
	ID                     string  `json:"id"`
	IntegrationID          string  `json:"integration_id"`
	SourceRealmID          string  `json:"source_realm_id"`
	IntegrationSlug        string  `json:"integration_slug"`
	IntegrationDisplayName string  `json:"integration_display_name"`
	RoleID                 string  `json:"role_id"`
	RoleName               string  `json:"role_name"`
	PrincipalUserID        string  `json:"principal_user_id"`
	ApprovedByUserID       *string `json:"approved_by_user_id"`
	ApprovedAt             string  `json:"approved_at"`
	LastUsedAt             *string `json:"last_used_at"`
	MintCount              int64   `json:"mint_count"`
}

// IntegrationCreate is the register body.
type IntegrationCreate struct {
	Slug        string `json:"slug"`
	DisplayName string `json:"display_name"`
	Description string `json:"description,omitempty"`
	HomepageURL string `json:"homepage_url,omitempty"`
	Listed      bool   `json:"listed,omitempty"`
}

// IntegrationPatch is the update body. Pointer fields signal "include";
// nil means "don't touch".
type IntegrationPatch struct {
	DisplayName *string `json:"display_name,omitempty"`
	Description *string `json:"description,omitempty"`
	HomepageURL *string `json:"homepage_url,omitempty"`
	Listed      *bool   `json:"listed,omitempty"`
}

// InstallRequest is the install body. RoleID MUST name a role whose
// assignable_to is exactly ["service"] (ADR-082 §7.1).
type InstallRequest struct {
	IntegrationID string `json:"integration_id"`
	RoleID        string `json:"role_id"`
}

// InstallResult is the install acknowledgment.
type InstallResult struct {
	ID              string `json:"id"`
	IntegrationID   string `json:"integration_id"`
	RoleID          string `json:"role_id"`
	RoleName        string `json:"role_name"`
	PrincipalUserID string `json:"principal_user_id"`
	Status          string `json:"status"`
}

// IntegrationMintRequest is the input to MintToken. APIKey is the SOURCE platform's raw
// platform_api key (never a user/session token). SourceOrgID is required and
// stamped into the token + target-org audit, but is caller-asserted (§7.6).
type IntegrationMintRequest struct {
	APIKey         string
	InstallationID string
	SourceOrgID    string
}

// IntegrationMintResult is the brokered token. There is NO refresh token: the token
// cannot be renewed, so re-mint when it nears expiry. ExpiresIn is a fixed
// 600 seconds (ADR-083 §4.3).
type IntegrationMintResult struct {
	AccessToken string `json:"access_token"`
	ExpiresIn   int64  `json:"expires_in"`
	TenantID    string `json:"tenant_id"`
	Role        string `json:"role"`
}

// IntegrationListPage / InstallationListPage are one raw page each (SPEC §7).
type IntegrationListPage struct {
	Items      []Integration `json:"items"`
	NextCursor *string       `json:"next_cursor"`
}
type InstallationListPage struct {
	Items      []Installation `json:"items"`
	NextCursor *string        `json:"next_cursor"`
}

// ListOpts are the optional pagination inputs shared by the list surfaces.
type ListOpts struct {
	Cursor string
	Limit  int
}

// Integration error sentinels. Callers branch with errors.Is.
var (
	ErrIntegrationNotFound  = errors.New("realmid: integration not found")
	ErrIntegrationSlugTaken = errors.New("realmid: integration slug already registered in realm")
	ErrAlreadyInstalled     = errors.New("realmid: integration already installed in org")
	ErrRoleNotServiceTyped  = errors.New(`realmid: role is not exactly ["service"]`)
	ErrInstallationNotFound = errors.New("realmid: installation not found")
	ErrInstallationRevoked  = errors.New("realmid: installation has been revoked")
	ErrRoleUnavailable      = errors.New("realmid: the installed role can no longer back an integration")
	ErrKeyClassMismatch     = errors.New("realmid: this grant requires a platform api key")
)

func mapIntegrationErr(err error) error {
	var re *RealmError
	if !errors.As(err, &re) {
		return err
	}
	// See specificCode: a registered code lands in re.Code and never in the
	// siblings, an unregistered one only in the siblings.
	switch specificCode(re) {
	case "integration_not_found":
		return errors.Join(ErrIntegrationNotFound, re)
	case "slug_taken":
		return errors.Join(ErrIntegrationSlugTaken, re)
	case "already_installed":
		return errors.Join(ErrAlreadyInstalled, re)
	case "role_not_service_typed", "role_not_installable":
		return errors.Join(ErrRoleNotServiceTyped, re)
	case "installation_not_found":
		return errors.Join(ErrInstallationNotFound, re)
	case "installation_revoked":
		return errors.Join(ErrInstallationRevoked, re)
	case "role_unavailable":
		return errors.Join(ErrRoleUnavailable, re)
	case "key_class_mismatch":
		return errors.Join(ErrKeyClassMismatch, re)
	}
	return re
}

// IntegrationsClient is realm.Integrations — both sides of the cross-realm
// integration surface plus the mint.
type IntegrationsClient struct {
	realm *Realm
}

func (c *IntegrationsClient) sourceBase() string {
	return "/platforms/" + url.PathEscape(c.realm.realmID) + "/integrations"
}
func (c *IntegrationsClient) targetBase(tenantID string) string {
	return "/tenants/" + url.PathEscape(tenantID) + "/integration-installations"
}

func listQuery(opts *ListOpts) map[string]string {
	q := map[string]string{}
	if opts != nil {
		if opts.Cursor != "" {
			q["cursor"] = opts.Cursor
		}
		if opts.Limit > 0 {
			q["limit"] = strconv.Itoa(opts.Limit)
		}
	}
	return q
}

// ---- source side ----

// Register publishes a new integration in the source realm.
func (c *IntegrationsClient) Register(ctx ctxpkg.Context, body IntegrationCreate) (*Integration, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var out Integration
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST", Path: c.sourceBase(), Bearer: tok, Body: body,
	}, &out); err != nil {
		return nil, mapIntegrationErr(err)
	}
	return &out, nil
}

// List returns one page of the source realm's published integrations.
func (c *IntegrationsClient) List(ctx ctxpkg.Context, opts *ListOpts) (*IntegrationListPage, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var page IntegrationListPage
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET", Path: c.sourceBase(), Bearer: tok, Query: listQuery(opts),
	}, &page); err != nil {
		return nil, mapIntegrationErr(err)
	}
	if page.Items == nil {
		page.Items = []Integration{}
	}
	return &page, nil
}

// Update patches an integration's display fields / listed flag.
func (c *IntegrationsClient) Update(ctx ctxpkg.Context, id string, patch IntegrationPatch) (*Integration, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var out Integration
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "PATCH", Path: c.sourceBase() + "/" + url.PathEscape(id), Bearer: tok, Body: patch,
	}, &out); err != nil {
		return nil, mapIntegrationErr(err)
	}
	return &out, nil
}

// Disable halts every mint (reversible via Enable).
func (c *IntegrationsClient) Disable(ctx ctxpkg.Context, id string) error {
	return c.toggle(ctx, id, "disable")
}

// Enable re-enables a disabled integration.
func (c *IntegrationsClient) Enable(ctx ctxpkg.Context, id string) error {
	return c.toggle(ctx, id, "enable")
}

func (c *IntegrationsClient) toggle(ctx ctxpkg.Context, id, verb string) error {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return err
	}
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST", Path: c.sourceBase() + "/" + url.PathEscape(id) + "/" + verb, Bearer: tok,
	}, nil); err != nil {
		return mapIntegrationErr(err)
	}
	return nil
}

// Remove permanently disables an integration (the source half of two-ended
// revocation). It is NOT a cascade delete — target orgs' inbound history
// survives (ADR-083 §9).
func (c *IntegrationsClient) Remove(ctx ctxpkg.Context, id string) error {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return err
	}
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "DELETE", Path: c.sourceBase() + "/" + url.PathEscape(id), Bearer: tok,
	}, nil); err != nil {
		return mapIntegrationErr(err)
	}
	return nil
}

// ---- target side ----

// Install admits a foreign integration into a target org. The role named by
// body.RoleID MUST be exactly ["service"] (ADR-082 §7.1) — otherwise
// ErrRoleNotServiceTyped.
func (c *IntegrationsClient) Install(ctx ctxpkg.Context, tenantID string, body InstallRequest) (*InstallResult, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var out InstallResult
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST", Path: c.targetBase(tenantID), Bearer: tok, Body: body,
	}, &out); err != nil {
		return nil, mapIntegrationErr(err)
	}
	return &out, nil
}

// ListInstallations returns one page of the target org's inbound access. A
// non-zero count after an ownership transfer is foreign access the new owner
// never approved (ADR-082 §7.4) — surface it.
func (c *IntegrationsClient) ListInstallations(ctx ctxpkg.Context, tenantID string, opts *ListOpts) (*InstallationListPage, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var page InstallationListPage
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET", Path: c.targetBase(tenantID), Bearer: tok, Query: listQuery(opts),
	}, &page); err != nil {
		return nil, mapIntegrationErr(err)
	}
	if page.Items == nil {
		page.Items = []Installation{}
	}
	return &page, nil
}

// Uninstall revokes an inbound edge (the target half of two-ended revocation).
// Future mints fail; live access tokens are NOT revoked (bounded by the 600 s
// TTL, ADR-083 §4.4).
func (c *IntegrationsClient) Uninstall(ctx ctxpkg.Context, tenantID, installationID string) error {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return err
	}
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "DELETE", Path: c.targetBase(tenantID) + "/" + url.PathEscape(installationID), Bearer: tok,
	}, nil); err != nil {
		return mapIntegrationErr(err)
	}
	return nil
}

// ---- mint ----

// MintToken mints a brokered target-realm access token against an installation,
// authenticated by the SOURCE platform's raw platform_api key (NOT a
// user/session token). It returns an access token only — no refresh — so
// re-mint as expiry nears. This is deliberately not wired into a token manager
// (§6.14): the token cannot refresh.
func (c *IntegrationsClient) MintToken(ctx ctxpkg.Context, req IntegrationMintRequest) (*IntegrationMintResult, error) {
	var out IntegrationMintResult
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/auth/login",
		// No Bearer — the raw api_key in the body IS the credential.
		Body: map[string]any{
			"grant_type":      "integration_installation",
			"api_key":         req.APIKey,
			"installation_id": req.InstallationID,
			"source_org_id":   req.SourceOrgID,
		},
	}, &out); err != nil {
		return nil, mapIntegrationErr(err)
	}
	return &out, nil
}

File: ts/src/integrations.ts
/**
 * Cross-realm integrations — `realm.integrations.*` (ADR-082 / ADR-083).
 *
 * A SOURCE platform publishes an integration; a TARGET org installs it,
 * admitting a `kind=service` principal into the org that holds a chosen
 * service-typed role; the source platform then MINTS short-lived target-realm
 * access tokens against the installation. GitHub-App-shaped. RI hosts no
 * consent screen — this surface IS the consent surface (ADR-083 §5).
 *
 * The SDK is per-realm: register/mint run on the SOURCE realm's client,
 * install/uninstall run on the TARGET realm's client. The source-side methods
 * take no platform id (baked in, like `realm.roles.*`).
 *
 * Server error codes surface on `RealmError.code`: `slug_taken` (409),
 * `already_installed` (409), `permissions_required` / `unknown_permission`
 * (400), `permissions_exceed_grantor` (403), `integration_not_found` /
 * `installation_not_found` (404), `installation_revoked` / `role_unavailable`
 * (403), `key_class_mismatch` (401). `role_not_service_typed` /
 * `role_not_installable` are retained in the taxonomy but DEAD — the issuer
 * has emitted neither since ADR-101 D7.
 */

import type { HttpClient } from "./http.js";

/** A source platform's published integration. */
export interface Integration {
  id: string;
  realm_id: string;
  slug: string;
  display_name: string;
  description: string;
  homepage_url: string;
  listed: boolean;
  disabled: boolean;
  created_at: string;
  updated_at: string;
  [k: string]: unknown;
}

/** One inbound edge in a target org's access list (ADR-083 §4.5). */
export interface Installation {
  id: string;
  integration_id: string;
  source_realm_id: string;
  integration_slug: string;
  integration_display_name: string;
  /**
   * The ADR-101 D7 stated grant — what the brokered principal may do. It
   * REPLACED `role_id`/`role_name`, which named a role and inherited whatever
   * that role happened to grant that day.
   */
  permissions: string[];
  principal_user_id: string;
  approved_by_user_id: string | null;
  approved_at: string;
  last_used_at: string | null;
  mint_count: number;
  [k: string]: unknown;
}

/** POST /platforms/{id}/integrations body. */
export interface IntegrationCreate {
  slug: string;
  displayName: string;
  description?: string;
  homepageUrl?: string;
  listed?: boolean;
}

/** PATCH /platforms/{id}/integrations/{iid} body. Omitted keys are untouched. */
export interface IntegrationPatch {
  displayName?: string;
  description?: string;
  homepageUrl?: string;
  listed?: boolean;
}

/**
 * POST /tenants/{id}/integration-installations body.
 *
 * `permissions` is the ADR-101 D7 STATED grant: the ADR-074 catalog
 * permissions this integration may exercise in the target org. It replaced
 * `roleId`, which named a role and silently inherited whatever that role
 * granted today.
 *
 * Required and non-empty — an install granting nothing can authorise no call,
 * and ADR-100's lesson is that an empty authority field acquires a meaning
 * nobody chose. Empty is `permissions_required`, not an install that enforces
 * nothing. Every entry must be a real catalog permission
 * (`unknown_permission`), and you cannot grant authority you do not hold
 * (`permissions_exceed_grantor`).
 */
export interface InstallRequest {
  integrationId: string;
  permissions: string[];
}

/** Install acknowledgment. */
export interface InstallResult {
  id: string;
  integration_id: string;
  permissions: string[];
  principal_user_id: string;
  status: string;
  [k: string]: unknown;
}

/**
 * Input to `mintToken`. `apiKey` is the SOURCE platform's raw `platform_api`
 * key (never a user/session token). `sourceOrgId` is required and stamped into
 * the token + target-org audit, but is caller-asserted (ADR-082 §7.6).
 */
export interface IntegrationMintRequest {
  apiKey: string;
  installationId: string;
  sourceOrgId: string;
}

/**
 * Brokered token. There is NO refresh token — the token cannot be renewed, so
 * re-mint as expiry nears. `expires_in` is a fixed 600 s (ADR-083 §4.3).
 */
export interface IntegrationMintResult {
  access_token: string;
  expires_in: number;
  tenant_id: string;
  role: string;
  [k: string]: unknown;
}

/** One page of integrations / installations (locked SPEC §7 envelope). */
export interface IntegrationListPage {
  items: Integration[];
  next_cursor: string | null;
}
export interface InstallationListPage {
  items: Installation[];
  next_cursor: string | null;
}

/** Optional pagination inputs shared by the list surfaces. */
export interface IntegrationListOpts {
  cursor?: string;
  limit?: number;
}

function integrationPage(raw: unknown): IntegrationListPage {
  const obj = (raw ?? {}) as Record<string, unknown>;
  const items = Array.isArray(obj["items"]) ? (obj["items"] as Integration[]) : [];
  const nc = obj["next_cursor"];
  return { items, next_cursor: typeof nc === "string" ? nc : null };
}
function installationPage(raw: unknown): InstallationListPage {
  const obj = (raw ?? {}) as Record<string, unknown>;
  const items = Array.isArray(obj["items"]) ? (obj["items"] as Installation[]) : [];
  const nc = obj["next_cursor"];
  return { items, next_cursor: typeof nc === "string" ? nc : null };
}

export class IntegrationsClient {
  constructor(
    private readonly http: HttpClient,
    private readonly realmId: string,
  ) {}

  private sourceBase(): string {
    return `/platforms/${encodeURIComponent(this.realmId)}/integrations`;
  }
  private targetBase(tenantId: string): string {
    return `/tenants/${encodeURIComponent(tenantId)}/integration-installations`;
  }

  // ---- source side ----

  /** POST /platforms/{id}/integrations — publish a new integration. */
  async register(body: IntegrationCreate): Promise<Integration> {
    const wire: Record<string, unknown> = { slug: body.slug, display_name: body.displayName };
    if (body.description !== undefined) wire["description"] = body.description;
    if (body.homepageUrl !== undefined) wire["homepage_url"] = body.homepageUrl;
    if (body.listed !== undefined) wire["listed"] = body.listed;
    return this.http.request<Integration>({ method: "POST", path: this.sourceBase(), body: wire });
  }

  /** GET /platforms/{id}/integrations — one page of published integrations. */
  async list(opts?: IntegrationListOpts): Promise<IntegrationListPage> {
    const raw = await this.http.request<unknown>({
      method: "GET",
      path: this.sourceBase(),
      query: { cursor: opts?.cursor, limit: opts?.limit },
    });
    return integrationPage(raw);
  }

  /** PATCH /platforms/{id}/integrations/{iid} — edit display fields / listed. */
  async update(id: string, patch: IntegrationPatch): Promise<Integration> {
    const wire: Record<string, unknown> = {};
    if (patch.displayName !== undefined) wire["display_name"] = patch.displayName;
    if (patch.description !== undefined) wire["description"] = patch.description;
    if (patch.homepageUrl !== undefined) wire["homepage_url"] = patch.homepageUrl;
    if (patch.listed !== undefined) wire["listed"] = patch.listed;
    return this.http.request<Integration>({
      method: "PATCH",
      path: `${this.sourceBase()}/${encodeURIComponent(id)}`,
      body: wire,
    });
  }

  /** POST …/{iid}/disable — reversible halt of every mint. */
  async disable(id: string): Promise<void> {
    await this.http.request<unknown>({
      method: "POST",
      path: `${this.sourceBase()}/${encodeURIComponent(id)}/disable`,
    });
  }

  /** POST …/{iid}/enable — re-enable a disabled integration. */
  async enable(id: string): Promise<void> {
    await this.http.request<unknown>({
      method: "POST",
      path: `${this.sourceBase()}/${encodeURIComponent(id)}/enable`,
    });
  }

  /**
   * DELETE /platforms/{id}/integrations/{iid} — permanent disable (the source
   * half of two-ended revocation). NOT a cascade delete; target orgs' inbound
   * history survives (ADR-083 §9).
   */
  async remove(id: string): Promise<void> {
    await this.http.request<unknown>({
      method: "DELETE",
      path: `${this.sourceBase()}/${encodeURIComponent(id)}`,
    });
  }

  // ---- target side ----

  /**
   * POST /tenants/{id}/integration-installations — admit a foreign integration,
   * granting it exactly the permissions `body.permissions` names (ADR-101 D7).
   */
  async install(tenantId: string, body: InstallRequest): Promise<InstallResult> {
    return this.http.request<InstallResult>({
      method: "POST",
      path: this.targetBase(tenantId),
      body: { integration_id: body.integrationId, permissions: body.permissions },
    });
  }

  /**
   * GET /tenants/{id}/integration-installations — the inbound-access list. A
   * non-zero count after an ownership transfer is foreign access the new owner
   * never approved (ADR-082 §7.4) — surface it.
   */
  async listInstallations(tenantId: string, opts?: IntegrationListOpts): Promise<InstallationListPage> {
    const raw = await this.http.request<unknown>({
      method: "GET",
      path: this.targetBase(tenantId),
      query: { cursor: opts?.cursor, limit: opts?.limit },
    });
    return installationPage(raw);
  }

  /**
   * DELETE /tenants/{id}/integration-installations/{iid} — revoke an inbound
   * edge. Future mints fail; live access tokens are NOT revoked (bounded by the
   * 600 s TTL, ADR-083 §4.4).
   */
  async uninstall(tenantId: string, installationId: string): Promise<void> {
    await this.http.request<unknown>({
      method: "DELETE",
      path: `${this.targetBase(tenantId)}/${encodeURIComponent(installationId)}`,
    });
  }

  // ---- mint ----

  /**
   * Mint a brokered target-realm access token against an installation,
   * authenticated by the SOURCE platform's raw `platform_api` key (NOT a
   * user/session token). Returns an access token only — no refresh — so re-mint
   * as expiry nears. `skipPlatformToken` keeps the SDK's own platform bearer
   * off this call: the raw `apiKey` in the body IS the credential.
   */
  async mintToken(req: IntegrationMintRequest): Promise<IntegrationMintResult> {
    return this.http.request<IntegrationMintResult>({
      method: "POST",
      path: "/auth/login",
      skipPlatformToken: true,
      body: {
        grant_type: "integration_installation",
        api_key: req.apiKey,
        installation_id: req.installationId,
        source_org_id: req.sourceOrgId,
      },
    });
  }
}
